### PR TITLE
Feature/ethers/account manager

### PR DESCRIPTION
**Pull request by jurajpiar**
## What

- re-add AccountManager module
- adapt it to using ethers

## Why

- part of:

## Refs

- [ticket](https://rsklabs.atlassian.net/browse/PP-490)
- [parent](https://rsklabs.atlassian.net/browse/PP-489)

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -54,6 +54,7 @@
   },
   "dependencies": {
     "@metamask/eth-sig-util": "^5.0.0",
+    "@nomicfoundation/hardhat-toolbox": "^2.0.0",
```
Comment (jurajpiar) [inline package.json:57]: remove

File: package.json [hunk 2]
```diff
@@ -45,7 +45,7 @@
     "prepack": "npmignore --auto",
     "prepare": "scripts/prepare",
     "prepublishOnly": "scripts/prepublishOnly",
-    "tdd": "npm run test -- --watch --watch-files src,test",
+    "tdd": "npm run test -- --watch --reporter mocha-clearscreen-reporter --watch-files src,test",
```
Comment (jurajpiar) [inline package.json:48]: remove

File: package.json [hunk 3]
```diff
@@ -83,16 +84,16 @@
     "eslint": "~8.25.0",
     "eslint-config-prettier": "^8.5.0",
     "eslint-plugin-mocha": "^10.1.0",
-    "ethereum-waffle": "^3.4.4",
+    "eslint-plugin-prettier": "^4.2.1",
     "husky": "^8.0.1",
     "lint-staged": "^13.0.3",
-    "mocha": "^10.0.0",
+    "mocha": "^9.0.0",
+    "mocha-clearscreen-reporter": "^1.0.1",
```
Comment (jurajpiar) [inline package.json:91]: remove

File: src/AccountManager.ts [hunk 4]
```diff
@@ -0,0 +1,134 @@
+import sigUtil, {
+  SignTypedDataVersion,
+  TypedMessage,
+} from '@metamask/eth-sig-util';
+import { providers, Signer, Wallet } from 'ethers';
+// import type { EnvelopingConfig } from './types/EnvelopingConfig';
```
Comment (jurajpiar) [inline src/AccountManager.ts:6]: remove

File: src/AccountManager.ts [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+  }
+
+  private async _signWithProvider<T>(
+    data: TypedMessage<EnvelopingMessageTypes>
+  ): Promise<T> {
+    if (!(this._provider as providers.JsonRpcProvider).send) {
+      throw new Error(`Not an RPC provider`);
+    }
+
+    return (await (this._provider as providers.JsonRpcBatchProvider).send(
```
Comment (jurajpiar) [inline src/AccountManager.ts:118]:
```suggestion
    return (await (this._provider as providers.JsonRpcBatchProvider).connect(this._signer).send(
```

Comment (jurajpiar) [reply src/AccountManager.ts:118]: Not sure about this id which is just a date. Can anyone clarify whether it is used for anything?

Screenshot 2022-11-01 at 2 05 23 pm

Comment (franciscotobar) [reply src/AccountManager.ts:118]: As far of my understanding. The Json RPC requests always needs an Id, which is meaningless for the blockchain but it could mean something to us. If they sent the date and its not use in other place, we can send anything.

File: src/typedRequestData.utils.ts [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+  verifier: string;
+  relayRequest:
+    | EnvelopingTypes.RelayRequestStruct
+    | EnvelopingTypes.DeployRequestStruct;
+  requestTypes: MessageTypeProperty[];
+};
+
+export const getEnvelopingRequestDataV4Field = ({
+  chainId,
+  verifier, // is this separated from the relay request by design?
```
Comment (jurajpiar) [inline src/typedRequestData.utils.ts:89]: When would we want to use this with verifier external to the request that is being passed as third arg?

File: src/typedRequestData.utils.ts [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+  relayRequest,
+  requestTypes,
+}: GetRequestDataFieldProps): TypedMessage<EnvelopingMessageTypes> => ({
+  types: {
+    EIP712Domain: eip712DomainType,
+    RelayRequest: requestTypes,
+    RelayData: relayDataType,
+  },
+  primaryType: 'RelayRequest', // what if it's deploy request?
```
Comment (jurajpiar) [inline src/typedRequestData.utils.ts:98]: Is this the correct primary type? Would have expected this to be `DeployRequest` for deploy requests...

File: test/types/AccountManager.test.ts [hunk 8]
```diff
@@ -0,0 +1,312 @@
...
+use(sinonChai);
+use(chaiAsPromised);
+
+const sandbox = createSandbox();
+
+function createRandomAddress() {
+  const randomHex = crypto.randomBytes(20).toString('hex');
+
+  return getAddress(`0x${randomHex}`);
+}
```
Comment (jurajpiar) [inline test/types/AccountManager.test.ts:21]: actually, this might be better using wallet.createWallet when I think about it...

File: test/types/AccountManager.test.ts [hunk 9]
```diff
@@ -0,0 +1,312 @@
...
+        };
+
+        accountManager = new AccountManager(stubAdmin, stubProvider, 1);
+      });
+
+      describe('addAccount', function () {
+        it('should add wallet clone to local accounts', function () {
+          const wallet = Wallet.createRandom();
+
+          // sandbox.stub(providers.Provider, 'isProvider').callsFake(() => true)
```
Comment (jurajpiar) [inline test/types/AccountManager.test.ts:115]: remove

File: src/typedRequestData.utils.ts (hunk 7 above)
Comment (franciscotobar) [reply src/typedRequestData.utils.ts:98]: The name was standardized but we could change it. If we are going to change it, we need to change in the contracts too, in the factories to be specific.

File: src/typedRequestData.utils.ts (hunk 6 above)
Comment (franciscotobar) [reply src/typedRequestData.utils.ts:89]: This verifier its not verifierContracts. This verifier is the address that we are trying to recover with the domain separator. It could be either the smartWallet address, or the smartWalletFactory.

File: src/AccountManager.ts (hunk 5 above)
Comment (jurajpiar) [reply src/AccountManager.ts:118]: just looked into the implementation of the jsonrpcprovider and it autoincrements it's own local id that begins with 42 lol.

File: package.json (hunk 1 above)
Comment (franciscotobar) [reply package.json:57]: Im not sure what do you mean with by remove, well I do but not pretty sure.

File: src/typedRequestData.utils.ts (hunk 6 above)
Comment (jurajpiar) [reply src/typedRequestData.utils.ts:89]: I don't understand your answer, so I'll ask it differently:
the verifier is taken from the relay request and then the whole request is passed in as well in the account manager https://github.com/rsksmart/rif-relay-client/pull/34/files#diff-54caa818924335d964cb20a8998a3476391b16f046153c048a6147986d157e7fL92. Why is it important to keep those separate? Are there any scenarios where the verifier would not be taken from the request.

File: src/typedRequestData.utils.ts (hunk 7 above)
Comment (jurajpiar) [reply src/typedRequestData.utils.ts:98]: I mean, we have `RelayRequest` and `DeployRequest` types. Yet, here we hardcode it to the latter, and I just wonder: is that the expected primary type for all and if so, then why?

File: src/AccountManager.ts [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+    }
+    this._accounts.push(wallet);
+  }
+
+  isDeployRequest({ request }: RelayOrDeployRequest): boolean {
+    return 'recoverer' in request;
+  }
+
+  async sign(relayRequest: RelayOrDeployRequest): Promise<string> {
+    const cloneRequest = { ...relayRequest }; // why does it have to be clone? don't seem to be mutated.
```
[the remark this answers is not in the document]
Comment (jurajpiar) [reply src/AccountManager.ts:52]: ok I'm removing it.

File: src/typedRequestData.utils.ts (hunk 6 above)
Comment (franciscotobar) [reply src/typedRequestData.utils.ts:89]: The verifier is not taken from the relay request. This verifier is built in, in the domain separator. During the process of the signature recovering from the EIP712, the domain separator is needed to recover the external signature from the EOA from the client, not the worker.

File: src/typedRequestData.utils.ts (hunk 7 above)
Comment (franciscotobar) [reply src/typedRequestData.utils.ts:98]: Yes its expected to be the RelayRequest for the two types of transactions. Why Im not sure why they kept the same name, it could be changed, but that change includes a change in the contracts. 

This primary type that is declared here, serves to identify which is the struct has inside a nested struct. 

`RelayRequest(address relayHub,address from,address to,address tokenContract,address recoverer,uint256 value,uint256 nonce,uint256 tokenAmount,uint256 tokenGas,uint256 index,bytes data,RelayData relayData)RelayData(uint256 gasPrice,address feesReceiver,address callForwarder,address callVerifier)`

This line of code is inside the SmartWallet and SmartWalletFactory. As you can see the RelayRequest has inside of it the RelayData, and if we want we could change those names.

File: src/typedRequestData.utils.ts (hunk 6 above)
Comment (jurajpiar) [reply src/typedRequestData.utils.ts:89]: [see this line](https://github.com/rsksmart/rif-relay-client/blob/2d3f1668afa088d1e17e49ca7fbd515d191b93fa/src/AccountManager.ts#L92)
Screenshot 2022-11-01 at 4 21 46 pm
Screenshot 2022-11-01 at 4 23 24 pm

File: src/AccountManager.ts [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+  addAccount(account: Wallet): void {
+    const wallet = new Wallet(account.privateKey, this._provider);
+    if (wallet.address !== account.address) {
+      throw new Error('invalid keypair');
+    }
+    this._accounts.push(wallet);
+  }
+
+  isDeployRequest({ request }: RelayOrDeployRequest): boolean {
+    return 'recoverer' in request;
```
Comment (franciscotobar) [inline src/AccountManager.ts:44]: I would use the index instead of the recoverer.

File: src/typedRequestData.utils.ts [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+  verifyingContract: string,
+  chainId: number
+): EIP712Domain {
+  return {
+    name: 'string name,string version',
+    version: '2',
+    chainId: chainId,
+    verifyingContract: verifyingContract,
+  };
+}
```
Comment (franciscotobar) [inline src/typedRequestData.utils.ts:65]:
```suggestion
function getDomainSeparator(
  verifyingContract: string,
  chainId: number
): EIP712Domain {
  return {
    name: domainSeparatorType.name,
    version: domainSeparatorType.version,
    chainId: chainId,
    verifyingContract: verifyingContract,
  };
}
```

Comment (franciscotobar) [reply src/typedRequestData.utils.ts:65]: The domain separator structure includes a name and version, but you are putting the structure and it must match with the one inside the contracts:

https://github.com/rsksmart/rif-relay-contracts/blob/master/contracts/factory/SmartWalletFactory.sol#L84
https://github.com/rsksmart/rif-relay-contracts/blob/master/contracts/smartwallet/SmartWallet.sol#L27

File: test/AccountManager.test.ts [hunk 13]
```diff
@@ -0,0 +1,292 @@
...
+            expectedMessage,
+            sandbox.match.any
+          );
+        });
+
+        it('should verify the request originator address against recovered typed signature', async function () {
+          sandbox
+            .stub(
+              accountManager as unknown as {
+                _recoverSignature: () => string;
```
Comment (franciscotobar) [inline test/AccountManager.test.ts:210]: should we use this approach for stubbing private keys?

File: test/AccountManager.test.ts [hunk 14]
```diff
@@ -0,0 +1,292 @@
...
+
+    describe('', function () {
+      let deployRequest: EnvelopingTypes.DeployRequestStruct;
+      let relayRequest: EnvelopingTypes.RelayRequestStruct;
+
+      let accountManager: AccountManager;
+      let stubProvider: providers.JsonRpcProvider;
+
+      beforeEach(function () {
+        const relayOrDeployRequest = {
```
Comment (franciscotobar) [inline test/AccountManager.test.ts:52]: if we are stubbing the methods, why do we need the whole object? I try just to use the necessary properties and for ce the conversion to the required object.

File: src/typedRequestData.utils.ts (hunk 12 above)
Comment (jurajpiar) [reply src/typedRequestData.utils.ts:65]: First, my bad for not checking the code I copied over from https://github.com/rsksmart/rif-relay-contracts/blob/master/types/EIP712/TypedRequestData.ts#L82 (I would not likely have written
`{ ..., chainId: chainId, verifyingContract: verifyingContract }`).
Second, I really don't know what you're trying to convey here, sorry. Please rephrase.

File: test/AccountManager.test.ts (hunk 14 above)
Comment (jurajpiar) [reply test/AccountManager.test.ts:52]: Could you please clarify as to why you prefer to use only the necessary objects for given test?
I'd argue that:
1. it's better to build your objects that do not require change once (which, fair enough, is not what I am doing here, but probably should)
2. here specifically, it allowed me to split and see what is actually different between relay and deploy requests
3. by default you want to pass a complete object and if you want test for a lack of a property you remove it
4. reduces the amount of constants you need to then always build your object

File: src/AccountManager.ts (hunk 11 above)
Comment (jurajpiar) [reply src/AccountManager.ts:44]: 1. explain
2. the original implementation uses recoverer

File: test/AccountManager.test.ts (hunk 14 above)
Comment (franciscotobar) [reply test/AccountManager.test.ts]: There is no specific reason. Just trying to follow some standard. Ill change my test to use the whole object object.

File: test/AccountManager.test.ts (hunk 13 above)
Comment (franciscotobar) [reply test/AccountManager.test.ts:206]: no alternative os stubbing privates, on the unit tests that I have done, the private method was execute since I did not stub it.

File: src/typedRequestData.utils.ts (hunk 12 above)
Comment (franciscotobar) [reply src/typedRequestData.utils.ts:65]: In your code the domain separator is wrong. It should be like this:

```suggestion
export function getDomainSeparator(
  verifyingContract: string,
  chainId: number
): EIP712Domain {
  return {
    name: 'RSK Enveloping Transaction',
    version: '2',
    chainId: chainId,
    verifyingContract: verifyingContract,
  };
}
```

File: src/AccountManager.ts (hunk 11 above)
Comment (franciscotobar) [reply src/AccountManager.ts:44]: The recoverer function is not implemented yet and it could be optional since we are using always the zero address, but the index cannot be optional at all since its necessary.

File: src/typedRequestData.utils.ts (hunk 12 above)
Comment (jurajpiar) [reply src/typedRequestData.utils.ts:65]: that's just repeating yourself.
but actually I do see now what you mean. I don't know how the name parameter managed to get changed like that...

Comment (jurajpiar) [reply src/typedRequestData.utils.ts:65]: done

File: test/AccountManager.test.ts (hunk 13 above)
Comment (jurajpiar) [reply test/AccountManager.test.ts:206]: ok then I don't know what the status of this comment is now....

File: test/AccountManager.test.ts (hunk 14 above)
Comment (jurajpiar) [reply test/AccountManager.test.ts]: ok so resolve this comment please if there is nothing to follow up on

Comment (jurajpiar) [reply test/AccountManager.test.ts]: btw I moved it above the main describe block

File: src/AccountManager.ts [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+    const data = getEnvelopingRequestDataV4Field({
+      chainId: this.chainId,
+      verifier: callForwarder,
+      relayRequest,
+      requestTypes: isDeploy ? deployRequestType : relayRequestType,
+    });
+
+    const wallet = this._accounts.find(
+      (account) => getAddress(account.address) === getAddress(from)
+    );
```
Comment (antomor) [inline src/AccountManager.ts:67]:
```suggestion
    const fromAddress = getAddress(from)
    const wallet = this._accounts.find(
      (account) => getAddress(account.address) === fromAddress
    );
```

File: src/AccountManager.ts [hunk 16]
```diff
@@ -0,0 +1,128 @@
...
+      } else {
+        signature = await this._signWithProvider(data, from);
+      }
+      rec = this._recoverSignature(data, signature);
+    } catch (error) {
+      throw new Error(
+        `Failed to sign relayed transaction for ${from}: ${error as string}`
+      );
+    }
+    if (getAddress(rec) !== getAddress(from)) {
```
Comment (antomor) [inline src/AccountManager.ts:83]:
```suggestion
    if (getAddress(rec) !== fromAddress) {
```

File: test/AccountManager.test.ts (hunk 13 above)
Comment (antomor) [reply test/AccountManager.test.ts:206]: @jurajpiar I think that the point here is: if I stub the private methods, how am I going to test them? Should we create unit tests for private methods? I'd say no unless strictly required, so we should probably execute the private methods without stubbing them, shouldn't we?

**Review by antomor — CHANGES_REQUESTED** [review]

File: test/AccountManager.test.ts (hunk 13 above)
Comment (jurajpiar) [reply test/AccountManager.test.ts:206]: not if the private method doesn't contribute to the test case

File: src/AccountManager.ts [hunk 17]
```diff
@@ -0,0 +1,128 @@
...
+  }
+
+  async sign(relayRequest: RelayOrDeployRequest): Promise<string> {
+    const {
+      relayData: { callForwarder },
+      request: { from },
+    } = relayRequest as typeof relayRequest & {
+      relayData: { callForwarder: string };
+      request: { from: string };
+    }; // this is here just to remove PromiseOrValue<string> type from the def string properties
```
[the remark this answers is not in the document]
Comment (jurajpiar) [reply src/AccountManager.ts:54]: because then we'd always have to either cast or convert the variable to string. The type exists for their (typechain) ethers v5 target to avoid multiple awaits in nested calls. This is not our case however.
But yea, I guess this casting may cause issue if in future this type changes. I'll just convert then.

[the remark this answers is not in the document]
Comment (jurajpiar) [reply src/AccountManager.ts]: done

File: src/AccountManager.ts (hunk 15 above)
Comment (jurajpiar) [reply src/AccountManager.ts]: done

File: src/AccountManager.ts (hunk 16 above)
Comment (jurajpiar) [reply src/AccountManager.ts]: done

File: test/AccountManager.test.ts (hunk 13 above)
Comment (antomor) [reply test/AccountManager.test.ts:206]: >Also another issue is that all this private method does is calling an dependency. Now, normally the dependency should be stubbed, but I was unable to stub sigUtil directly, so I wrapped it in a method. Should you find a better solution I will be more than happy to change it.

got it, if you couldn't stub the dependency, it makes sense to me

**Review by antomor — APPROVED** [review]
LG
